rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java

```diff
@@ -1,69 +1,301 @@
 
-        // Finished syncing
-        syncEventsHandler.onCompletedSyncing();
+        clearPeerInfo(peerId);
+        if (suitablePeers.isEmpty()) {
+            syncEventsHandler.stopSyncing();
+            return;
+        }
+        // if this peer has another different message pending then its restored to the stack
+        Long messageId = messagesByPeers.remove(peerId);
+        if (messageId != null) {
+            resetChunkAndHeader(peerId, pendingBodyResponses.remove(messageId).header);
+        }
+        startDownloading(getInactivePeers());
+    }
+
+    private void resetChunkAndHeader(NodeID peerId, BlockHeader header) {
+        Integer chunkNumber = chunksBeingDownloaded.remove(peerId);
+        pendingHeaders.get(chunkNumber).push(header);
+        Integer segmentNumber = segmentsBeingDownloaded.remove(peerId);
+        chunksBySegment.get(segmentNumber).push(chunkNumber);
+    }
+
+    private void clearPeerInfo(NodeID peerId) {
+        suitablePeers.remove(peerId);
+        timeElapsedByPeer.remove(peerId);
+    }
+
+    private Optional<BlockHeader> updateHeadersAndChunks(NodeID peerId, Integer currentChunk) {
+        Stack<BlockHeader> headers = pendingHeaders.get(currentChunk);
+        if (!headers.empty()) {
+            return Optional.of(headers.pop());
+        }
+
+        Optional<BlockHeader> header = tryFindBlockHeader(peerId);
+        if (!header.isPresent()){
+            chunksBeingDownloaded.remove(peerId);
+            segmentsBeingDownloaded.remove(peerId);
+            messagesByPeers.remove(peerId);
+        }
+
+        return header;
+    }
+
+    private Optional<BlockHeader> tryFindBlockHeader(NodeID peerId) {
+        Integer segmentNumber = segmentByNode.get(peerId);
+        // we start from the last chunk that can be downloaded
+        while (segmentNumber >= 0){
+            Stack<Integer> chunks = chunksBySegment.get(segmentNumber);
+            // if the segment stack is empty then continue to next segment
+            if (!chunks.empty()) {
+                Integer chunkNumber = chunks.pop();
+                Stack<BlockHeader> headers = pendingHeaders.get(chunkNumber);
+                if (!headers.empty()) {
+                    chunksBeingDownloaded.put(peerId, chunkNumber);
+                    segmentsBeingDownloaded.put(peerId, segmentNumber);
+                    return Optional.of(headers.pop());
+                } else {
+                    // log something we are in trouble
+                }
+            }
+            segmentNumber--;
+        }
+        return Optional.empty();
     }
 
     @Override
     public void onEnter() {
-        requestBody();
+        initializeSegments();
+        startDownloading(suitablePeers);
+    }
+
+    private void startDownloading(List<NodeID> peers) {
+        for (int i = 0; i < peers.size(); i++){
+            NodeID peerId = peers.get(i);
+            Optional<BlockHeader> nextHeader = tryFindBlockHeader(peerId);
```

Comment:
try to reuse code from `tryRequestNextBody`

Reply:
Improved code